### PR TITLE
Generate script functions

### PR DESCRIPTION
**Pull request by danhedron**
To fix:
- [x] Strange man appears in kurma during intro
- [x] Fading doesn't work correctly
- [x] Door doesn't rotate
- [x] Script stops after re-entering kurma in cutscene
- [x] Double kurma in the introduction mission due to a problem `00d8`
- [x] Misty never enters the car
- [x] Misty's radar blip is at the origin
- [x] Object replacement needs restoring
- [ ] Crashes when loading Luigi's Girls save. (is pickup collected 0x214)
- [ ] Crashes when you approach the dock. (garage contains needed car 0x3d4)

---

This moves all of the bind() calls into a single GTA3 Module, so that the implementation file only contains the actual functions themselves.

The included script can re-generate the file from the opcode interface descriptions

Implemented instructions to migrate
- [x] 0x001
- [x] 0x002
- [x] 0x004
- [x] 0x005
- [x] 0x006
- [x] 0x008
- [x] 0x009
- [x] 0x00C
- [x] 0x00D
- [x] 0x015
- [x] 0x018
- [x] 0x019
- [x] 0x01A
- [x] 0x01B
- [x] 0x020
- [x] 0x028
- [x] 0x029
- [x] 0x02A
- [x] 0x038
- [x] 0x039
- [x] 0x042
- [x] 0x04D
- [x] 0x04E
- [x] 0x04F
- [x] 0x050
- [x] 0x051
- [x] 0x0053
- [x] 0x0055
- [x] 0x0056
- [x] 0x0057
- [x] 0x060
- [x] 0x061
- [x] 0x084
- [x] 0x086
- [x] 0x009A
- [x] 0x009B
- [x] 0x00A1
- [x] 0x00A5
- [x] 0x00A6
- [x] 0x00AA
- [x] 0x00BA
- [x] 0x00BC
- [x] 0x00BE
- [x] 0x00BF
- [x] 0x00C0
- [x] 0x0D6
- [x] 0x0D7
- [x] 0x0D8
- [x] 0x00D9
- [x] 0x00DA
- [x] 0x00DB
- [x] 0x00DC
- [x] 0x00DE
- [x] 0x00DF
- [x] 0x00E0
- [x] 0x00E1
- [x] 0x00E5
- [x] 0x00E6
- [x] 0x00E8
- [x] 0x00E9
- [x] 0x00EB
- [x] 0x00ED
- [x] 0x00EF
- [x] 0x00F6
- [x] 0x00F7
- [x] 0x00F9
- [x] 0x0100
- [x] 0x0108
- [x] 0x0111
- [ ] 0x0112
- [x] 0x0118
- [ ] 0x0119
- [x] 0x0121
- [ ] 0x0122
- [x] 0x0126
- [x] 0x0129
- [x] 0x0137
- [x] 0x014B
- [x] 0x014C
- [x] 0x0152
- [x] 0x0159
- [x] 0x015A
- [x] 0x015C
- [x] 0x015F
- [x] 0x0160
- [x] 0x0164
- [x] 0x0169
- [x] 0x016A
- [x] 0x016B
- [x] 0x016E
- [x] 0x0171
- [x] 0x0173
- [x] 0x0175
- [x] 0x0176
- [x] 0x0177
- [x] 0x0180
- [x] 0x0186
- [x] 0x0187
- [x] 0x018A
- [x] 0x018B
- [x] 0x0199
- [x] 0x019C
- [x] 0x019E
- [x] 0x01A0
- [x] 0x01A8
- [x] 0x01AA
- [x] 0x01AD
- [x] 0x01B2
- [x] 0x01B4
- [x] 0x01B6
- [x] 0x01BD
- [x] 0x01C1
- [ ] 0x01C7
- [x] 0x01C8
- [x] 0x01D3
- [x] 0x01D4
- [x] 0x01D5
- [x] 0x01DF
- [x] 0x01E3
- [x] 0x01E7
- [x] 0x01E8
- [x] 0x01F0
- [ ] 0x01F3
- [x] 0x01F4
- [x] 0x01F5
- [x] 0x01F6
- [x] 0x0204
- [x] 0x0211
- [x] 0x0213
- [x] 0x0214
- [x] 0x0215
- [x] 0x0219
- [x] 0x021C
- [x] 0x0227
- [x] 0x0229
- [x] 0x022A
- [x] 0x022B
- [x] 0x0239
- [x] 0x023C
- [x] 0x023D
- [x] 0x0244
- [ ] 0x0248
- [x] 0x0255
- [x] 0x0256
- [x] 0x0293
- [x] 0x029B
- [ ] 0x029C
- [x] 0x029F
- [x] 0x02A3
- [x] 0x02A7
- [x] 0x02A8
- [ ] 0x02B3
- [ ] 0x02BF
- [x] 0x2CD
- [x] 0x02DE
- [x] 0x02E3
- [x] 0x02E4
- [x] 0x02E5
- [x] 0x02E6
- [x] 0x02E7
- [x] 0x02E8
- [x] 0x02E9
- [x] 0x02EA
- [x] 0x02EB
- [x] 0x02ED
- [x] 0x02F3
- [x] 0x02F4
- [x] 0x02F5
- [x] 0x02FB
- [x] 0x030C
- [x] 0x030D
- [x] 0x0314
- [x] 0x0318
- [ ] 0x0320
- [x] 0x0324
- [x] 0x0329
- [x] 0x032B
- [x] 0x0339
- [x] 0x033E
- [x] 0x0340
- [x] 0x0346
- [x] 0x034D
- [x] 0x034F
- [x] 0x0352
- [x] 0x0354
- [x] 0x0355
- [ ] 0x0363
- [x] 0x0394
- [x] 0x0395
- [x] 0x3A4
- [ ] 0x03B6
- [ ] 0x03C6
- [x] 0x03CF
- [x] 0x03D0
- [x] 0x03D1
- [x] 0x03D2
- [x] 0x03D4
- [x] 0x03D5
- [ ] 0x03D9
- [x] 0x03DC
- [x] 0x03E1
- [x] 0x03E5
- [x] 0x03E6
- [ ] 0x03EC
- [ ] 0x03EE
- [ ] 0x03F3
- [ ] 0x03F7
- [x] 0x0408
- [x] 0x040A
- [x] 0x417
- [x] 0x041D
- [x] 0x0422
- [x] 0x042C
- [x] 0x043C
- [x] 0x043D
- [x] 0x0442
- [x] 0x0443
- [ ] 0x0445
- [x] 0x0448
- [x] 0x044D

### CODE REVIEW

File: rwengine/CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,66 @@
...
+	src/script/modules/GameModule.cpp
+	src/script/modules/ObjectModule.cpp
+	src/script/modules/VMModule.cpp
+	src/script/SCMFile.cpp
+	src/script/ScriptDisassembly.cpp
+	src/script/ScriptMachine.cpp
+	src/script/ScriptModule.cpp
+	src/script/ScriptTypes.cpp
+	src/TextureArchive.cpp
+)
```
Comment (JayFoxRox) [inline rwengine/CMakeLists.txt:135]: does this drop the headers? why?

Also this should probably be a seperate branch imho. A PR could probably be fast-tracked + build/makefile changes often cause regressions for some people, so we should attempt to isolate those.

File: scripts/generate_script_functions.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            m = re.search("bindFunction\(\s?0x([0-9A-Fa-f]+), (-?\d+), ([0-9A-Za-z_<>]+)", line)
+            if m is not None:
+                opcode = int(m.group(1), 16)
+                name = m.group(3)
+                opcode_names[opcode] = name
+            m = re.search("bindUnimplemented\(\s?0x([0-9A-Fa-f]+), (-?\d+), ([0-9A-Za-z_<>]+)", line)
+            if m is not None:
+                opcode = int(m.group(1), 16)
+                name = m.group(3)
+                unimplemented_names[opcode] = name
```
Comment (JayFoxRox) [inline scripts/generate_script_functions.py]: Your current solution works for now but I think in the long term searching for implemented opcodes should be done differently or.. not at all.

Given that larger SCM.ini changes are highly unlikely in the future we could generate all function names from that.
We'd then generate all binds (there wouldn't be anything unimplemented) and stubs for all functions (with type from SCM.ini).
If we want to bind functions to templates we'd do it near the implementation instead of in the bind.

We can then simply turn stubs into functions as we see fit.
In the _highly_ unlikely event that there is a major rename / refactoring in the SCM.ini we'd just generate a short-script to search for the stubs which need a rename.

Having all stubs ready for implementation also reduces the burden to actually implement a feature. In the long run this will save time. (0x300 opcodes \* optimistic 10 seconds to change from bindUnimplemented to a bindFunction and adding a stub with argument parsing:  7680 seconds ~> 2 hours)

This means, that as a one-time-solution we need a tool to rename our existing script functions into what SCM.ini would expect AND we need a handful of rules to rename opcodes (such as "/=" becoming something like: "divide_by" etc.)
We should also have pseudo data-types such as LocalInteger and GlobalInteger we can use as arguments to the stub.
Information which is not yet present in SCM.ini should be added by us, possibly in the form as an overlay SCM.ini, patch or seperate file. That could contain information that a certain argument is a local/global/garage/crane/certain type of pickup/.. .
We could, however, also just cast in the actual implementation functions.

File: rwengine/include/script/ScriptModule.hpp [hunk 3]
```diff
@@ -25,14 +52,35 @@ class ScriptModule
...
 	);
-	
+
+	template <class Tfunc>
+	void bind(ScriptFunctionID id,
+			  int args,
+			  Tfunc function)
+	{
+		functions.insert({id, {
+							  [=](const ScriptArguments& args) {
```
Comment (JayFoxRox) [inline rwengine/include/script/ScriptModule.hpp]: "tabs are better than spaces" ;)

File: rwengine/src/script/modules/GTA3ModuleImpl.inl [hunk 4]
```diff
@@ -28,8 +31,13 @@ void opcode_0001(const ScriptArguments& args, const ScriptInt time) {
 */
 void opcode_0002(const ScriptArguments& args, const ScriptLabel arg1) {
 	RW_UNIMPLEMENTED_OPCODE(0x0002);
-	RW_UNUSED(arg1);
-	RW_UNUSED(args);
+	auto thread = args.getThread();
+	if (arg1 < 0) {
```
Comment (JayFoxRox) [inline rwengine/src/script/modules/GTA3ModuleImpl.inl:44]: shouldn't this be handled by the ScriptLabel parser?
[or a function of the ScriptLabel type]

Comment (danhedron) [inline rwengine/src/script/modules/GTA3ModuleImpl.inl:44]: Yes.

File: rwengine/include/script/ScriptTypes.hpp [hunk 5]
```diff
@@ -140,6 +300,52 @@ template<> GameObject* ScriptArguments::getObject<PickupObject>(unsigned int arg
...
+template<> glm::u8vec4 ScriptArguments::getParameter<glm::u8vec4>(unsigned int arg) const;
+
+template<> float& ScriptArguments::getParameterRef<float>(unsigned int arg) const;
+template<> int& ScriptArguments::getParameterRef<int>(unsigned int arg) const;
+
+///////////////////////////////////////////////////////////
+// Script Object specialisations
+template <>
+ScriptObjectType<VehicleObject> ScriptArguments::getScriptObject(
+    unsigned int arg) const;
```
Comment (darkf) [inline rwengine/include/script/ScriptTypes.hpp]: Can these arguments go on the same line (it's not 1980 anymore if you're worried about line length.)

File: rwengine/include/script/ScriptTypes.hpp [hunk 6]
```diff
@@ -140,6 +311,52 @@ template<> GameObject* ScriptArguments::getObject<PickupObject>(unsigned int arg
...
+ScriptObjectType<BlipData> ScriptArguments::getScriptObject(
+    unsigned int arg) const;
+template <>
+ScriptObjectType<VehicleGenerator> ScriptArguments::getScriptObject(
+    unsigned int arg) const;
+template <>
+ScriptObjectType<GarageInfo> ScriptArguments::getScriptObject(
+    unsigned int arg) const;
+
+/// @todo remove this temporary placeholder
```
Comment (darkf) [inline rwengine/include/script/ScriptTypes.hpp]: Can this be resolved?

File: rwengine/src/script/ScriptMachine.cpp [hunk 7]
```diff
@@ -148,12 +148,18 @@ void ScriptMachine::executeThread(SCMThread &t, int msPassed)
...
-				printf(" %08x", a.integerValue());
+				if (a.type == SCMType::TString) {
+					printf(" %01x:% 8s", a.type, a.string);
+				}
+				else {
+					printf(" %01x:%08x", a.type, a.integerValue());
+				}
 			}
 			printf("\n");
+			fflush(0);
```
Comment (darkf) [inline rwengine/src/script/ScriptMachine.cpp]: a) This seems unnecessary since stdout/stderr are usually line-buffered.

b) Even if it was, you should at least write `fflush(stdout);` or whatever (be explicit)

File: rwengine/src/script/ScriptMachine.cpp [hunk 8]
```diff
@@ -148,12 +148,18 @@ void ScriptMachine::executeThread(SCMThread &t, int msPassed)
 #if RW_SCRIPT_DEBUG
 		if (strcmp(t.name, "EIGHT") == 0)
```
Comment (darkf) [inline rwengine/src/script/ScriptMachine.cpp:149]: std::string

File: rwengine/src/script/ScriptTypes.cpp [hunk 9]
```diff
@@ -95,3 +95,151 @@ template<> GameObject* ScriptArguments::getObject< VehicleObject >(unsigned int
 	auto it = getWorld()->objects.find( gameObjectID );
 	return (it == getWorld()->objects.end())? nullptr : it->second;
 }*/
+
+template<> float ScriptArguments::getParameter<float>(unsigned int arg) const
+{
+	return getParameters()[arg].realValue();
```
Comment (darkf) [inline rwengine/src/script/ScriptTypes.cpp]: We should be using at() here (or wrapping std::vector) instead of operator[] to get bounds-checking.

File: rwengine/include/script/ScriptTypes.hpp [hunk 10]
```diff
@@ -27,6 +29,145 @@ typedef uint16_t SCMOpcode;
...
+struct VehicleGenerator;
+struct BlipData;
+struct GarageInfo;
+
+using ScriptVehicleGenerator = ScriptObjectType<VehicleGenerator>;
+using ScriptBlip = ScriptObjectType<BlipData>;
+using ScriptGarage = ScriptObjectType<GarageInfo>;
+
+/* Not Implemented Yet */
+using ScriptSound = ScriptObjectType<int>;
```
Comment (darkf) [inline rwengine/include/script/ScriptTypes.hpp]: Why are these all int vs a more explicit type?

File: rwengine/src/script/ScriptTypes.cpp [hunk 11]
```diff
@@ -95,3 +95,144 @@ template<> GameObject* ScriptArguments::getObject< VehicleObject >(unsigned int
...
+template<> glm::u8vec4 ScriptArguments::getParameter<glm::u8vec4>(unsigned int arg) const
+{
+	RW_CHECK(getParameters().size() > arg+3, "Argument list too short to get u8vec4");
+	return glm::u8vec4(getParameter<int>(arg), getParameter<int>(arg+1), getParameter<int>(arg+2), getParameter<int>(arg+3));
+}
+
+template<> float& ScriptArguments::getParameterRef<float>(unsigned int arg) const
+{
+	auto& param = (*this)[arg];
+	return *((float*)param.handleValue());
```
Comment (darkf) [inline rwengine/src/script/ScriptTypes.cpp]: a) `reinterpret_cast` please

b) This will be wrong if `sizeof int != sizeof float`

c) This may violate strict aliasing rules anyway, you should use memcpy to copy the int into a float. (It's optimized out, but avoids UB).

File: rwengine/src/script/ScriptTypes.cpp [hunk 12]
```diff
@@ -95,3 +95,144 @@ template<> GameObject* ScriptArguments::getObject< VehicleObject >(unsigned int
...
+}
+
+template<> int& ScriptArguments::getParameterRef<int>(unsigned int arg) const
+{
+	auto& param = (*this)[arg];
+	return *((int*)param.handleValue());
```
Comment (darkf) [inline rwengine/src/script/ScriptTypes.cpp]: Same for this.

File: rwengine/include/script/ScriptTypes.hpp (hunk 10 above)
Comment (danhedron) [inline rwengine/include/script/ScriptTypes.hpp]: They're not implemented yet, they'll be replaced with a proper type once the functionality for those are implemented.

Comment (JayFoxRox) [inline rwengine/include/script/ScriptTypes.hpp]: The comment should include TODO or FIXME then, so that we can keep track of it more easily.

Otherwise we might stop using one of those types in the future and we'll forget about these declarations. The TODO / FIXME would remind us of incomplete code.

File: rwengine/include/script/ScriptFunctions.hpp [hunk 13]
```diff
@@ -13,6 +13,8 @@
  * among many script modules and opcodes.
  */
 namespace script {
+const ScriptVec3 kSpawnOffset = ScriptVec3(0.f, 0.f, 1.f);
```
Comment (JayFoxRox) [inline rwengine/include/script/ScriptFunctions.hpp]: I feel this commit should be in another PR if possible.

Please attempt to make this change a straight refactor.
It will be hard to review it otherwise.
